File: apps/pi/pi/services/retrievers/pg_store/webhook.py
# SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
# SPDX-License-Identifier: LicenseRef-Plane-Commercial
#
# Licensed under the Plane Commercial License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# https://plane.so/legals/eula
#
# DO NOT remove or modify this notice.
# NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.

from typing import Optional

from sqlalchemy import desc
from sqlalchemy import select
from sqlmodel.ext.asyncio.session import AsyncSession

from pi import logger
from pi.app.models.github_webhook import GitHubWebhook

log = logger.getChild(__name__)


async def get_last_processed_commit(session: AsyncSession, repo_name: str, branch_name: str) -> Optional[str]:
    """
    Get the last successfully processed commit ID for a repository.
    """
    query = (
        select(GitHubWebhook)
        .where(GitHubWebhook.source == repo_name)  # type: ignore[var-annotated,arg-type]
        .where(GitHubWebhook.branch_name == branch_name)  # type: ignore[var-annotated,arg-type]
        .where(GitHubWebhook.processed)  # type: ignore[var-annotated,arg-type]
        .order_by(desc(GitHubWebhook.created_at))  # type: ignore[arg-type]
        .limit(1)
    )

    result = await session.execute(query)
    webhook = result.scalar_one_or_none()

    if webhook:
        log.info(f"Last processed commit for {repo_name}: {webhook.commit_id}")
        return webhook.commit_id

    log.info(f"No processed commits found for {repo_name}")
    return None


async def create_webhook_record(
    session: AsyncSession,
    commit_id: str,
    repo_name: str,
    branch_name: str,
    processed: bool = False,
    files_processed: Optional[int] = None,
    error_message: Optional[str] = None,
) -> GitHubWebhook:
    """
    Create a new webhook record in the database.
    """
    webhook = GitHubWebhook(
        commit_id=commit_id,
        source=repo_name,
        branch_name=branch_name,
        processed=processed,
        files_processed=files_processed,
        error_message=error_message,
    )

    session.add(webhook)
    await session.commit()
    await session.refresh(webhook)

    log.info(f"Created webhook record for {repo_name}: {commit_id} (processed: {processed})")
    return webhook


async def update_webhook_record(
    session: AsyncSession, webhook: GitHubWebhook, processed: bool, files_processed: Optional[int] = None, error_message: Optional[str] = None
) -> GitHubWebhook:
    """
    Update an existing webhook record.
    """
    webhook.processed = processed
    if files_processed is not None:
        webhook.files_processed = files_processed
    if error_message is not None:
        webhook.error_message = error_message

    session.add(webhook)
    await session.commit()
    await session.refresh(webhook)

    log.info(f"Updated webhook record for {webhook.source}: {webhook.commit_id} (processed: {processed})")
    return webhook


async def get_webhook_by_commit(session: AsyncSession, commit_id: str, repo_name: str, branch_name: str):
    """
    Get a webhook record by commit ID and repository name.
    """
    query = (
        select(GitHubWebhook)
        .where(GitHubWebhook.commit_id == commit_id)  # type: ignore[var-annotated,arg-type]
        .where(GitHubWebhook.source == repo_name)  # type: ignore[var-annotated,arg-type]
        .where(GitHubWebhook.branch_name == branch_name)  # type: ignore[var-annotated,arg-type]
    )

    result = await session.execute(query)
    return result.scalar_one_or_none()

File: packages/etl/src/gitlab/types/webhook.ts
/**
 * SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
 * SPDX-License-Identifier: LicenseRef-Plane-Commercial
 *
 * Licensed under the Plane Commercial License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://plane.so/legals/eula
 *
 * DO NOT remove or modify this notice.
 * NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.
 */

import type { GitlabIssue, GitlabLabel, GitlabMergeRequest, GitlabNote, GitlabProject, GitlabUser } from "./common";

// Base webhook event type
export interface GitlabWebhookEvent {
  object_kind: string;
  event_type?: string;
  user: GitlabUser;
  project: GitlabProject;
  repository: {
    name: string;
    url: string;
    description: string;
    homepage: string;
  };
  isEnterprise?: boolean;
}

// Push event
export interface GitlabPushEvent extends GitlabWebhookEvent {
  object_kind: "push";
  before: string;
  after: string;
  ref: string;
  checkout_sha: string;
  commits: {
    id: string;
    message: string;
    title: string;
    timestamp: string;
    url: string;
    author: {
      name: string;
      email: string;
    };
    added: string[];
    modified: string[];
    removed: string[];
  }[];
  total_commits_count: number;
}

// Issue event
export interface GitlabIssueEvent extends GitlabWebhookEvent {
  object_kind: "issue" | "work_item";
  event_type: "issue" | "work_item";
  object_attributes: GitlabIssue & {
    action: GitlabIssueWebhookActions;
    url: string;
    type: "Issue" | "Task" | "Incident";
  };
  labels: GitlabLabel[];
  changes: {
    [key: string]: {
      previous: any;
      current: any;
    };
  };
  assignees: GitlabUser[];
}

// Merge request event
export interface GitlabMergeRequestEvent extends GitlabWebhookEvent {
  object_kind: "merge_request";
  object_attributes: GitlabMergeRequest & {
    action: string;
    url: string;
    source: GitlabProject;
    target: GitlabProject;
    last_commit: {
      id: string;
      message: string;
      timestamp: string;
      url: string;
      author: {
        name: string;
        email: string;
      };
    };
  };
  labels: GitlabLabel[];
  changes: {
    [key: string]: {
      previous: any;
      current: any;
    };
  };
}

// Note event
export interface GitlabNoteEvent extends GitlabWebhookEvent {
  object_kind: "note";
  event_type: "note";
  object_attributes: GitlabNote & {
    description: string;
    url: string;
    action: GitlabNoteWebhookActions;
  };
  issue: Omit<GitlabIssue, "labels"> & {
    labels: GitlabLabel[];
  };
  merge_request?: GitlabMergeRequest;
  snippet?: {
    id: number;
    title: string;
    content: string;
    author_id: number;
    project_id: number;
    created_at: string;
    updated_at: string;
    file_name: string;
    type: string;
    visibility_level: number;
  };
  commit?: {
    id: string;
    message: string;
    timestamp: string;
    url: string;
    author: {
      name: string;
      email: string;
    };
  };
}

export enum GitlabIssueWebhookActions {
  OPEN = "open",
  CLOSE = "close",
  REOPEN = "reopen",
  UPDATE = "update",
}

export enum GitlabNoteWebhookActions {
  CREATE = "create",
  UPDATE = "update",
}
